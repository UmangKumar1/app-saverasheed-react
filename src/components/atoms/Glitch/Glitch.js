import React from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes } from 'react-emotion'
import { times, join, pipe, concat, append } from 'ramda'

const { string } = PropTypes

const random = num => Math.floor(Math.random() * Math.floor(num))

const createAnim = pipe(
  num =>
    times(idx => {
      const percentage = (idx / num) * 20
      const top = `${random(100)}%`
      const bottom = `${random(100)}%`
      return `
${percentage}% {
  clip-path: polygon(0% ${top}, 100% ${top}, 100% ${bottom}, 0% ${bottom});
}
`
    }, num),
  append(
    `
20%, 100% {
  clip-path: polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%);
  opacity: 0;
}
`,
  ),
  join('\n'),
)

const anim1 = createAnim(20)
const anim2 = createAnim(20)

const noiseAnim1 = keyframes`${anim1}`
const noiseAnim2 = keyframes`${anim2}`

const GlitchRoot = styled('div')({
  '&&': {
    width: 'auto',
    position: 'relative',
  },
})

const PseudoGlitch = styled('div')({
  // content: 'attr(data-text)',
  position: 'absolute',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%',
  overflow: 'hidden',
  clipPath: 'polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%)',
})

const GlitchBefore = styled(PseudoGlitch)({
  left: 2,
  textShadow: '-1px 0 red',
  animation: `${noiseAnim2} 5s infinite 1s linear alternate-reverse`,
})
const GlitchAfter = styled(PseudoGlitch)({
  left: -2,
  textShadow: '-1px 0 blue',
  animation: `${noiseAnim1} 5s infinite linear alternate-reverse`,
})

const Glitch = ({ children }) => (
  <GlitchRoot>
    <GlitchBefore>{children}</GlitchBefore>
    {children}
    <GlitchAfter>{children}</GlitchAfter>
  </GlitchRoot>
)

Glitch.propTypes = {
  children: string,
}

export default Glitch
