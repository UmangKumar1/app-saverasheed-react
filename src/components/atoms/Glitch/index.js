export { default, default as Glitch } from './Glitch'
