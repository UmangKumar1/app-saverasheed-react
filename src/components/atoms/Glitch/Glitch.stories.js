import React from 'react'
import { storiesOf } from '@storybook/react'
import Glitch from '.'

storiesOf('Glitch', module)
  .add('with text', () => <Glitch>Save Rasheed</Glitch>)
  .add('with image', () => (
    <Glitch>
      <img src="https://s3.amazonaws.com/uifaces/faces/twitter/millinet/128.jpg" />
    </Glitch>
  ))
