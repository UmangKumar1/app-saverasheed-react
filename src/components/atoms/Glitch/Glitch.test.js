import React from 'react'
import { shallow } from 'enzyme'
import Glitch from '.'

const wrap = (props = {}) => shallow(<Glitch {...props} />)

describe('Glitch', () => {
  it('renders children when passed in', () => {
    const wrapper = wrap({ children: 'Glitch' })
    expect(wrapper).toMatchSnapshot()
  })
})
