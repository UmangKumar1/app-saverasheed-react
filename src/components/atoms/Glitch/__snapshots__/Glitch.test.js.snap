// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glitch renders children when passed in 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Glitch>
    Glitch
  </Glitch>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Styled(div)>
          Glitch
        </Styled(div)>,
        "Glitch",
        <Styled(div)>
          Glitch
        </Styled(div)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Glitch",
        },
        "ref": null,
        "rendered": "Glitch",
        "type": [Function],
      },
      "Glitch",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Glitch",
        },
        "ref": null,
        "rendered": "Glitch",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Styled(div)>
            Glitch
          </Styled(div)>,
          "Glitch",
          <Styled(div)>
            Glitch
          </Styled(div)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Glitch",
          },
          "ref": null,
          "rendered": "Glitch",
          "type": [Function],
        },
        "Glitch",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Glitch",
          },
          "ref": null,
          "rendered": "Glitch",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
